pyro/distributions/polynomial.py

```diff
@@ -0,0 +1,139 @@
+from __future__ import absolute_import, division, print_function
+
+import math
+
+import torch
+import torch.nn as nn
+from torch.distributions import constraints
+
+from pyro.distributions.torch_transform import TransformModule
+from pyro.distributions.util import copy_docs_from
+
+
+@copy_docs_from(TransformModule)
+class PolynomialFlow(TransformModule):
+    """
+    An autoregressive normalizing flow as described in Jaini et al. (2019) using the element-wise transformation
+
+        :math:`y_n = c_n + \\int^{x_n}_0\\sum^K_{k=1}\\left(\\sum^R_{r=0}a^{(n)}_{r,k}u^r\\right)du`
+
+    where :math:`x_n` is the :math:`n`th input, :math:`y_n` is the :math:`n`th output, and :math:`c_n\\in\\mathbb{R}`,
+    :math:`\\left{a^{(n)}_{r,k}\\in\\mathbb{R}\\right}` are learnable parameters that are the output of an
+    autoregressive NN inputting :math:`x_{\\prec n}={x_1,x_2,\\ldots,x_{n-1}}`.
+
+    Together with `TransformedDistribution` this provides a way to create richer variational approximations.
+
+    Example usage:
+
+    >>> from pyro.nn import AutoRegressiveNN
+    >>> input_dim = 10
+    >>> count_degree = 4
+    >>> count_sum = 3
+    >>> base_dist = dist.Normal(torch.zeros(input_dim), torch.ones(input_dim))
+    >>> arn = AutoRegressiveNN(input_dim, [input_dim*10], param_dims=[(count_degree + 1)*count_sum])
+    >>> flow = PolynomialFlow(arn, input_dim=input_dim, count_degree=count_degree, count_sum=count_sum)
+    >>> pyro.module("my_flow", flow)  # doctest: +SKIP
+    >>> flow_dist = dist.TransformedDistribution(base_dist, [flow])
+    >>> flow_dist.sample()  # doctest: +SKIP
+        tensor([-0.4071, -0.5030,  0.7924, -0.2366, -0.2387, -0.1417,  0.0868,
+                0.1389, -0.4629,  0.0986])
+
+    The inverse of this transform does not possess an analytical solution and is left unimplemented. However,
+    the inverse is cached when the forward operation is called during sampling, and so samples drawn using
+    polynomial flow can be scored.
+
+    :param autoregressive_nn: an autoregressive neural network whose forward call returns a tensor of real-valued
+        numbers of size (batch_size, (count_degree+1)*count_sum, input_dim)
+    :type autoregressive_nn: nn.Module
+    :param count_degree: The degree of the polynomial to use for each element-wise transformation.
+    :type count_degree: int
+    :param count_sum: The number of polynomials to sum in each element-wise transformation.
+    :type count_sum: int
+
+    References:
+
+    Sum-of-squares polynomial flow. [arXiv:1905.02325]
+    Priyank Jaini, Kira A. Shelby, Yaoliang Yu
+    """
+
+    domain = constraints.real
+    codomain = constraints.real
+    bijective = True
+    event_dim = 1
+    autoregressive = True
+
+    def __init__(self, autoregressive_nn, input_dim, count_degree, count_sum):
```

Comment:
@stefanwebb could you please clarify how `count_degree` etc. correspond to the notation in the paper (r, k, etc.)? this will help me review the polynomial construction

Reply:
`r = count_degree`
`k = count_sum`

Another important detail: how the author's proposed to calculate the forward operation/gradient wasn't made explicit in the paper. I chose to expand the squared term in Eq (8) and integrate each power of `u`, which I think is the most numerically stable.

Reply:
And the Fundamental Theorem of Calculus for the Jacobian of course